torch/_inductor/codegen/wrapper.py

```diff
@@ -497,14 +514,10 @@ def generate_extern_kernel_out(self, output_view, codegen_reference, args, kerne
         self.writeline(f"{kernel}({', '.join(args)})")
 
     def generate_user_defined_triton_kernel(self, kernel_name, grid, configs, args):
-        assert len(grid) != 0
-        if len(grid) == 1:
-            grid = f"{grid[0]}"
-        else:
-            from torch._higher_order_ops.triton_kernel_wrap import grid_fn_code
-
-            grid, code = grid_fn_code(kernel_name, configs, grid)
-            self.header.splice(code, strip=True)
+        grid, code = user_defined_kernel_grid_fn_code(kernel_name, configs, grid)
+        # Must happen after free symbols are already codegened
+        with self.prefix.indent():
+            self.prefix.splice(code)
```

Comment:
Looking for feedback here. I moved the grid function to the prefix (rather than the header) so that it can access the free symbols from dynamic shapes. This happens to work because prefix contains the free symbols. Is there a better solution or is there a way for me to assert that free symbols are already generated?

Reply:
Because the grid function is a function (and not some inlined block of code) would all the symbols defined in the outer scope, including the ones defined *after* this function, not be visible in the inner scope of this function's body? E.g., this works:

```python
def fn(a):
  return a * b

b = 123
c = fn(456)
```

The main thing is that `fn` is *called before* the `b` is defined. But otherwise it's fine that `b` is defined after `fn`.

Reply:
I didn't know this is how python semantics worked.. C++ certainly does not work this way

Reply:
Depending on how we decide to deal with codegening the grid in the AOTInductor's C++ wrapper codegen, maybe we could opt for relying on this feature of Python.